### ISSUE TITLE
Since 1.0.0, an escaped non-utf8 query parameter raises with webmock

### ISSUE THREAD

**Issue by Earlopain** [excon/excon#863]
Caused by https://github.com/excon/excon/pull/857

```rb
require "excon"
require "webmock"

include WebMock::API
WebMock.enable!

param = 'aäoöuü'.encode('iso-8859-1')
param = CGI.escape(param)
stub_request(:get, "https://example.com/?#{param}").to_return(body: "abc")

Excon.get("https://example.com/?#{param}")
```

This is only the case when using it with webmock, just with excon itself and no mocking it passes. The params are escaped, however I believe excon unescapes them here, https://github.com/excon/excon/blob/33a071d992f04350c12918a9b259154df35c5332/lib/excon/utils.rb#L77. This produces an UTF-8 string that is not valid in its encoding. This is then passed along to webmock, where certain operations (like regexp matching) fail.

**Comment by geemus** [comment]
@Earlopain I think #865 should probably fix the problem. Would you be able to double check my work and your error case to see if it's now doing what you expect? Thanks!

**Comment by Earlopain** [comment]
Hm, I don't think that is doing much. I tried reproducing this without webmock and I think that's pretty simple since there is buildin mock functionality:

```rb
require "excon"

Excon.stub({:method => :get}) do |params|
  pp params[:query]
  { body: "OK", status: 200 }
end

connection = Excon.new('https://example.com', mock: true)

param = 'aäoöuü'.encode(Encoding::ISO8859_1)
param = CGI.escape(param).encode(Encoding::UTF_8)
connection.request(method: :get, query: param)
```

This prints `{"a\xE4o\xF6u\xFC" => nil}` but I would expect `{"a%E4o%F6u%FC" => nil}`, as passed to the function. If you try to call `params[:query].match?("foo")` it will raise. This is a bit of a weird case, but generally I would expect these to not be touched. Does that make sense?

Note that the passed string is already utf8 and valid, the unescaping is what creates the broken string.

**Comment by Earlopain** [comment]
Hm, rack/rails chokes on this:

```rb
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "example.org"
  config.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    match '*path', to: "test#index", via: :all
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def index
    pp params
    render plain: "Home"
  end
end

require "minitest/autorun"
require "rack/test"

class BugTest < ActiveSupport::TestCase
  include Rack::Test::Methods

  def test_returns_success
    param = 'aäoöuü'.encode('iso-8859-1')
    param = CGI.escape(param)
    get "/#{param}"
  end

  private
    def app
      Rails.application
    end
end
```

```
I, [2024-10-29T10:58:47.883467 #59775]  INFO -- : -----------------------------
I, [2024-10-29T10:58:47.883639 #59775]  INFO -- : BugTest: test_returns_success
I, [2024-10-29T10:58:47.883725 #59775]  INFO -- : -----------------------------
I, [2024-10-29T10:58:47.911418 #59775]  INFO -- : Started GET "/a%E4o%F6u%FC" for 127.0.0.1 at 2024-10-29 10:58:47 +0100
F, [2024-10-29T10:58:47.915811 #59775] FATAL -- :   
ActionController::BadRequest (Invalid path parameters: Invalid encoding for parameter: a�o�u�):

Causes:
Rack::QueryParser::InvalidParameterError (Invalid encoding for parameter: a�o�u�)
```

TBH, not sure what would be correct here. I have very little stake in this, I only noticed some test failures from webmock when doing a PR there: https://github.com/bblimke/webmock/pull/1074

**Comment by geemus** [comment]
Thanks for the repro case, that helps (I was a bit unsure of the case with webmock because I have less comfort/familiarity there).

Yeah, I was borrowing in part from Rack's handling of this in the hopes that it did the right thing. It's helpful to know it isn't, because then I can reset my expectations about whether I'm just borrowing incorrectly or it just doesn't work there.

Overall the parsing in excon for the mocks here is with the goal of making it so that whether you declare the mock with a string or a hash, you can also call it with either and it will still work. Previously if you define the mock with a string query, you had to call it with a string query (hash wouldn't match). So I was trying to reconcile that. I'll ponder some more and see what can be done.

**Comment by Earlopain** [comment]
I believe that change is good. I don't understand it entirely and would just remove any attempts at escaping but it's good for the initial report.

There's an small issue in webmock itself where it now double-encodes the values but that is hardly a fault in your library: https://github.com/bblimke/webmock/blob/9ff63ac7c845a75278e43236e0128e197f597462/lib/webmock/http_lib_adapters/excon_adapter.rb#L103

Interestingly, this isn't a problem right now but I can't really say why that would be. I'd expect it to have come up from the beginning but it didn't. Either way, probably something that should be changed in webmock itself and not here.

I'm out of my depth here, would you mind waiting a bit? I would ask about it in the webmock PR.

**Comment by geemus** [comment]
@Earlopain sure, I'm not in a hurry. I'd like to fix it, but suspect it won't impact too many people in the interim. I feel somewhat out of my depth around encodings too, so I appreciate your efforts to help regardless.

In terms of escaping, it's complicated. It would be great to just not escape/unescape, but that would break other expectations. In particular, I expect query stuff in a string to be escaped, but query stuff in a hash should NOT be escaped. Then, if escaping of the hash is necessary, the library will do that when building the query. Otherwise you can end up with double-escaping.

That being said, I'm not super familiar with webmock, but I wouldn't expect that to cause double-escaping. In particular, the goal with the changes in excon was that hash queries wouldn't be escaped. That method appears to take a hash query (which shouldn't be escaped) and convert it to a string (which should be escaped). Could you elaborate on how/where you are seeing the double escaping play out?

**Comment by Earlopain** [comment]
The approach in your current PR does fewer unescaping than previously:
```rb
input = "x=ab%20c&z=%27Stop%21%27%20said%20Fred%20m"
pr_parse_query_string(input)
=> {"x"=>"ab%20c", "z"=>"%27Stop%21%27%20said%20Fred%20m"}
master_parse_query_string(input)
=> {"x"=>"ab c", "z"=>"'Stop!' said Fred m"}
```

It comes percent-encoded to webmock, it percent-encodes them again. I can't tell which would be correct here.

**Comment by geemus** [comment]
Thanks, you are correct. I was missed that the gsub was happening but not actually being used (I think an `!` should fix that). I just pushed a change to that and the tests that I think bring it inline with expectations/intentions.

**Comment by geemus** [comment]
I've confirmed that the update results in master and this PR matching (unlike the previous attempt).

**Comment by geemus** [comment]
Ok, I think I was overthinking/overdoing it. I've simplified #869 so that instead of trying to do escaping, it just splits based on `&/;` and `:`. That should get the original goal (normalizing so that if you use a hash to define the mock but a string to call it, or vice versa, it still matches). I think I was thinking about the real case (where escaping comes into play). Especially since this is all internal for the sake of matching stubs, the encoding wasn't really gaining us much even if it was working (and it wasn't working).

@Earlopain thanks again for taking time to report this and help me with it, especially since it wasn't a direct issue for you. I really appreciate the assistance.

**Comment by geemus** [comment]
Released with fix in v1.1.1
